File: src/main/java/io/pivotal/pal/tracker/JdbcTimeEntryRepository.java
package io.pivotal.pal.tracker;

import static java.sql.Statement.NO_GENERATED_KEYS;
import static java.sql.Statement.RETURN_GENERATED_KEYS;

import java.sql.Date;
import java.sql.PreparedStatement;
import java.util.List;
import javax.sql.DataSource;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.ResultSetExtractor;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;

public class JdbcTimeEntryRepository implements TimeEntryRepository {

  private JdbcTemplate jdbcTemplate;

  public JdbcTimeEntryRepository(DataSource dataSource) {
    this.jdbcTemplate = new JdbcTemplate(dataSource);
  }

  @Override
  public TimeEntry create(TimeEntry timeEntry) {
    KeyHolder keyHolder = new GeneratedKeyHolder();

    jdbcTemplate.update(connection -> {
      PreparedStatement statement = connection.prepareStatement(
          "INSERT INTO time_entries (project_id, user_id, date, hours) " +
              "VALUES (?, ?, ?, ?)",
          RETURN_GENERATED_KEYS
      );

      statement.setLong(1, timeEntry.getProjectId());
      statement.setLong(2, timeEntry.getUserId());
      statement.setDate(3, Date.valueOf(timeEntry.getDate()));
      statement.setInt(4, timeEntry.getHours());

      return statement;
    }, keyHolder);
    return find(keyHolder.getKey().longValue());
  }

  @Override
  public TimeEntry find(long id) {
    return jdbcTemplate.query("SELECT * FROM time_entries WHERE id=?",
        new Object[] {id},
        extractor);
  }

  @Override
  public List<TimeEntry> list() {
    return jdbcTemplate.query("SELECT * FROM time_entries", rowMapper);
  }

  @Override
  public TimeEntry update(long id, TimeEntry timeEntry) {
    jdbcTemplate.update(connection -> {
      PreparedStatement statement = connection.prepareStatement(
          "UPDATE time_entries SET project_id = ?, user_id = ?, date = ?, hours = ? WHERE id= ? ",
          RETURN_GENERATED_KEYS
      );
      statement.setLong(1, timeEntry.getProjectId());
      statement.setLong(2, timeEntry.getUserId());
      statement.setDate(3, Date.valueOf(timeEntry.getDate()));
      statement.setInt(4, timeEntry.getHours());
      statement.setLong(5, id);


      return statement;
    });
    return find(id);
  }

  @Override
  public TimeEntry delete(long id) {
    TimeEntry timeEntry = find(id);
    jdbcTemplate.update(connection -> {
      PreparedStatement statement = connection.prepareStatement(
          "DELETE FROM time_entries WHERE id=?", NO_GENERATED_KEYS
      );

      statement.setLong(1, id);
      return statement;
    });
    return timeEntry;
  }

  private final RowMapper<TimeEntry> rowMapper = (rs, index) -> new TimeEntry(
      rs.getLong("id"),
      rs.getLong("project_id"),
      rs.getLong("user_id"),
      rs.getDate("date").toLocalDate(),
      rs.getInt("hours")
  );


  private final ResultSetExtractor<TimeEntry> extractor =
      (rs) -> rs.next() ? rowMapper.mapRow(rs, 1) : null;
}

File: src/main/java/io/pivotal/pal/tracker/JdbcTimeEntryRepository.kt
package io.pivotal.pal.tracker

import com.mysql.cj.jdbc.MysqlDataSource
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.PreparedStatementCreator
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.support.GeneratedKeyHolder
import java.sql.PreparedStatement
import java.util.*
import java.sql.Statement.RETURN_GENERATED_KEYS
import java.sql.Time
import java.time.LocalDate


class JdbcTimeEntryRepository(dataSource: MysqlDataSource): TimeEntryRepository {
    private var jdbcTemplate : JdbcTemplate

    init{
        this.jdbcTemplate = JdbcTemplate(dataSource)
    }

    override fun create(timeEntry: TimeEntry): TimeEntry {
        val keyHolder = GeneratedKeyHolder()
        val psc = PreparedStatementCreator {connection ->
            val statement: PreparedStatement = connection.prepareStatement(
                    "INSERT INTO time_entries (project_id, user_id, date, hours) " + "VALUES (?, ?, ?, ?)",
                    RETURN_GENERATED_KEYS
            )
            statement.setLong(1, timeEntry.projectId)
            statement.setLong(2, timeEntry.userId)
            statement.setDate(3, java.sql.Date.valueOf(timeEntry.date))
            statement.setInt(4, timeEntry.hours)
            statement
        }

        jdbcTemplate.update(psc, keyHolder)

        //find the new entry
        return find(keyHolder.key.toLong()) as TimeEntry
    }

    override fun find(id: Long): TimeEntry? {
        val timeEntries: List<TimeEntry> = jdbcTemplate.query("select * from time_entries where id =" + id, rowMapper)
        return if (timeEntries.isEmpty()) null else timeEntries.first()
    }

    override fun list(): List<TimeEntry>{
        return jdbcTemplate.query("select * from time_entries", rowMapper)
    }

    override fun update(id: Long, timeEntry: TimeEntry): TimeEntry{
        val psc = PreparedStatementCreator {connection ->
            val statement: PreparedStatement = connection.prepareStatement(
                    "UPDATE time_entries SET project_id = ?, user_id = ?, date = ?, hours = ? WHERE id = ?;" ,
                    RETURN_GENERATED_KEYS
            )
            statement.setLong(1, timeEntry.projectId)
            statement.setLong(2, timeEntry.userId)
            statement.setDate(3, java.sql.Date.valueOf(timeEntry.date))
            statement.setInt(4, timeEntry.hours)
            statement.setLong(5, id)
            statement
        }

        jdbcTemplate.update(psc)
        return find(id) as TimeEntry

    }

    override fun delete(id: Long): TimeEntry{
        val timeEntry = find(id) as TimeEntry

        jdbcTemplate.update({connection ->
            val statement: PreparedStatement = connection.prepareStatement(
                    "DELETE FROM time_entries WHERE id = " + id ,
                    RETURN_GENERATED_KEYS
            )
            statement
        })
        return timeEntry
    }

    private val rowMapper: RowMapper<TimeEntry> = RowMapper { resultSet, _ -> TimeEntry(resultSet.getLong(1),
            resultSet.getLong(2), resultSet.getLong(3), resultSet.getDate(4).toLocalDate(), resultSet.getInt(5))}

}